### PR TITLE
Add initial project files and dependencies

### PR DESCRIPTION
**Pull request by walteck**
## Description

Initial commit of project files and dependencies for th egithub permission manager.

## Context

This introduces the code to develop a GitHub App that allows a trusted list of users to request and approve permission elevation to "Owner" for the organisation

## Type of changes

- [ ] Refactoring (non-breaking change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [X] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [X] I have followed the code style of the project
- [X] I have added tests to cover my changes
- [X] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [X] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+        github_permission_manager.post_comment_on_issue(payload, comment_body)
+
+        mock_post.assert_called_once_with(
+            'https://api.github.com/repos/org/repo/issues/1/comments',
+            headers=github_permission_manager.auth_headers,
+            json={"body": comment_body}
+        )
+
+        assert mock_post.called
```
Comment (amaanibn-nasar1-nhs) [inline src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py:161]: the `GithubPermissionsManager` can take a requests module at instantiation so can't we just give it a mock module rather than mocking post

File: src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+        github_permission_manager.make_owner_on_github(payload, user)
+
+        mock_put.assert_called_once_with(
+            'https://api.github.com/orgs/test-org/memberships/test-user',
+            headers=github_permission_manager.auth_headers,
+            json={"role": "admin"}
+        )
+
+        assert mock_response.status_code == 200
```
Comment (amaanibn-nasar1-nhs) [inline src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py:182]: same as above

File: src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            f'https://api.github.com/orgs/{org}/teams/{team}',
+            headers=github_permission_manager.auth_headers
+        )
+
+        mock_get.assert_any_call(
+            f'https://api.github.com/teams/1/memberships/{user}',
+            headers=github_permission_manager.auth_headers
+        )
+
+        assert result == True
```
Comment (amaanibn-nasar1-nhs) [inline src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py:211]: same as above

File: src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+def test_handle_issue_comment(github_permission_manager):
+    payload = {
+        'action': 'created',
+        'comment': {'user': {'login': 'user'}, 'body': 'approve'},
+        'repository': {'owner': {'login': 'org'}}
+    }
+    with patch.object(github_permission_manager, '_is_user_eligible_for_elevation', return_value=True), \
+        patch.object(github_permission_manager, '_handle_approval_comment') as mock_handle:
+        github_permission_manager.handle_issue_comment(json.dumps(payload))
+        mock_handle.assert_called_once()
```
Comment (amaanibn-nasar1-nhs) [inline src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py:233]: I'm a bit confused what these two tests are doing, it's mocking two functions and then asserting that one of them is called?

File: src/github_permission_manager_demotion/handler.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    def get_all_org_owners(self, organization):
+        response = requests.get(
+            f"https://api.github.com/orgs/{organization}/members?role=admin",
+            headers=self.auth_headers
+        )
+        print(f"Request to get org owners: {response.status_code}")
+        org_owners = []
+        for owner in response.json():
+            print(f"Owner: {owner['login']}")
+            org_owners.append(owner['login'])
```
Comment (amaanibn-nasar1-nhs) [inline src/github_permission_manager_demotion/handler.py:74]: can we use a list comprehension here and then print the list afterwards?
I think it might be slightly neater but happy to leave as is

File: src/github_permission_manager_demotion/handler.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+            # Update DynamoDB
+            table.update_item(
+                Key={'user': user},
+                UpdateExpression="set #status = :s, demoted_at = :t",
+                ExpressionAttributeNames={'#status': 'status'},
+                ExpressionAttributeValues={
+                    ':s': 'demoted',
+                    ':t': datetime.now().isoformat()
+                }
+            )
```
Comment (amaanibn-nasar1-nhs) [inline src/github_permission_manager_demotion/handler.py:118]: can we flip this if statement as well please
`if user not in org_owners:` - log something and then return

File: src/github_permission_manager_demotion/handler.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        organization = event.get('organization')
+        user = event.get('user')
+        repository = event.get('repository')
+        issue_number = event.get('issue_number')
+        org_owners = self.get_all_org_owners(organization)
+        if user in org_owners:
+            if is_last_org_owner(org_owners, user):
+                self.post_comment_on_issue(repository, issue_number, "User is the last owner - therefore will not be demoted")
+                self.close_issue(repository, issue_number)
+                return
```
Comment (amaanibn-nasar1-nhs) [inline src/github_permission_manager_demotion/handler.py:101]: I feel like maybe there should be some more alarms bells on this, maybe something that posts to teams?

**Review by amaanibn-nasar1-nhs — COMMENTED** [review]
some comments and queries

File: src/github_permission_manager_webhook/handler.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+
+def request_is_to_elevate_access(issue):
+    keywords = ['request', 'elevate', 'elevation']
+    return any(keyword in issue.get('title', '') or keyword in issue.get('body', '') for keyword in keywords)
+
+def comment_contains_approval(comment):
+    return 'approve' in comment['body'] or '👍' in comment['body']
+
+def approving_own_request(user, original_requestor):
+    return user == original_requestor
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:22]: done

File: src/github_permission_manager_webhook/handler.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+from datetime import datetime
+from common.shared_functions import GitHubAuth, DEFAULT_REGION
+
+# Utility functions
+def get_headers_from_event(event):
+    return event.get('headers', {}) or event.get('Headers', {})
+
+def request_is_to_elevate_access(issue):
+    keywords = ['request', 'elevate', 'elevation']
+    return any(keyword in issue.get('title', '') or keyword in issue.get('body', '') for keyword in keywords)
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:16]: I've updated to use a regex and to ignorecase. However potentially in the future we might want to consider revisiting this approach.

File: src/github_permission_manager_webhook/handler.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+
+def comment_contains_approval(comment):
+    return 'approve' in comment['body'] or '👍' in comment['body']
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:19]: I've updated to use a regex and to ignorecase. However potentially in the future we might want to consider revisiting this approach.

File: src/github_permission_manager_webhook/handler.py [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+
+    def post_comment_on_issue(self, payload, comment_body):
+        print(f"Posting comment: {comment_body}")
+        data = {"body": comment_body}
+        response = self.requests.post(
+            f"https://api.github.com/repos/{payload['repository']['full_name']}/issues/{payload['issue']['number']}/comments",
+            headers=self.auth_headers,
+            json=data
+        )
+        print(f"Comment posted: {response.status_code}")
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:71]: I've added error handling but we should decide, in a future story, whether failure to post a comment should result in failure of the process, for example does the comment form part of the audit of the process?

File: src/github_permission_manager_webhook/handler.py [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+            if self._is_user_eligible_for_elevation(user, payload):
+                self._process_elevation_request(payload, issue, user)
+            else:
+                self._notify_ineligible_user(payload, user)
+
+    def _is_elevation_request(self, payload):
+        return payload['action'] == 'opened' and request_is_to_elevate_access(payload['issue'])
+
+    def _is_user_eligible_for_elevation(self, user, payload):
+        return self.is_team_member(user, payload['repository']['owner']['login'], 'can-escalate-to-become-an-owner')
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:148]: Done

File: src/github_permission_manager_webhook/handler.py [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+        return payload.encode('utf-8') if isinstance(payload, str) else payload
+
+    def _generate_signature(self, payload):
+        hash_object = hmac.new(self.webhook_secret.encode('utf-8'), msg=payload, digestmod=hashlib.sha256)
+        return "sha256=" + hash_object.hexdigest()
+
+    def _parse_payload(self, payload):
+        return json.loads(payload) if isinstance(payload, str) else payload
+
+    # Issue handling methods
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:133]: # assuming you mean the comment
have removed.

File: src/github_permission_manager_webhook/handler.py [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+        if response.status_code == 200:
+            print(f"User {username} is a member of the team.")
+            return True
+        elif response.status_code == 404:
+            print(f"User {username} is not a member of the team.")
+            return False
+        print(f"Error checking membership: {response.status_code}")
+        return False
+
+    # Request validation methods
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:110]: removed

File: src/github_permission_manager_webhook/handler.py [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+            f"https://api.github.com/orgs/{payload['organization']['login']}/memberships/{user}",
+            headers=self.auth_headers,
+            json=data
+        )
+        if response is None:
+            print("No response received from GitHub API")
+        else:
+            print(f"{'Promoted' if response.status_code == 200 else 'Failed to promote'} user to Owner: {response.status_code}")
+
+    # Team membership methods
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:85]: removed

File: src/github_permission_manager_webhook/handler.py [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+        table = self.dynamodb.Table('GithubElevationRequests')
+        table.put_item(Item={
+            'user': user,
+            'issue_number': issue['number'],
+            'repo': payload['repository']['full_name'],
+            'status': 'pending',
+            'requested_at': datetime.now().isoformat()
+        })
+
+    # Comment handling methods
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:167]: removed

File: src/github_permission_manager_webhook/handler.py [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+            user = comment['user']['login']
+            if self._is_comment_from_bot(user):
+                return
+            if not self._is_user_eligible_for_elevation(user, payload):
+                self.post_comment_on_issue(payload, f"@{user} has commented on the elevation request but is not a member of the elevators team.")
+                return
+            if comment_contains_approval(comment):
+                self._handle_approval_comment(payload, user)
+            else:
+                self.post_comment_on_issue(payload, f"@{user} has commented but not approved the elevation.")
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:181]: updated.

File: src/github_permission_manager_webhook/handler.py [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+
+    def _is_comment_from_bot(self, user):
+        return user == 'elevatemetoowner[bot]'
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:184]: changed.

File: src/github_permission_manager_webhook/handler.py [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+
+    def _handle_approval_comment(self, payload, user):
+        original_requestor = payload['issue']['user']['login']
+        self.post_comment_on_issue(payload, f"@{user} has approved the elevation for @{original_requestor}.")
+        if approving_own_request(user, original_requestor):
+            self.post_comment_on_issue(payload, f"@{user} cannot approve own requests.")
+            return
+        self.promote_user_to_owner(payload, original_requestor)
+
+    # User promotion methods
```
[the remark this answers is not in the document]
Comment (walteck) [reply src/github_permission_manager_webhook/handler.py:194]: removed.

File: src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py (hunk 1 above)
Comment (walteck) [reply src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py:161]: Changed this.

File: src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py (hunk 2 above)
Comment (walteck) [reply src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py:182]: changed

File: src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py (hunk 3 above)
Comment (walteck) [reply src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py:211]: changed

File: src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py (hunk 4 above)
Comment (walteck) [reply src/github_permission_manager_webhook/tests/test_handler_github_permission_manager.py:233]: Its testing the flow through the handle_issue() function, mocking the return value as true for the _is_user_eligible_for_elevation() call. I've added a test for ineligible users as well, where the return value is false.

File: src/github_permission_manager_demotion/handler.py (hunk 5 above)
Comment (walteck) [reply src/github_permission_manager_demotion/handler.py:74]: I've converted to a list comprehension and removed the print statement as it was intended for debugging.

File: src/github_permission_manager_demotion/handler.py (hunk 6 above)
Comment (walteck) [reply src/github_permission_manager_demotion/handler.py:118]: changed

File: src/github_permission_manager_demotion/handler.py (hunk 7 above)
Comment (walteck) [reply src/github_permission_manager_demotion/handler.py:101]: I'd suggest this is something for a future story to define what should happen here. For the initial work and running in a test organisation I'd say this is OK, but before taking any further we should work through the edge cases.

**Review by amaanibn-nasar1-nhs — APPROVED** [review]
